File: bot/telegram.py
"""
Telegram Notifications
"""

import os
import aiohttp


class TelegramNotifier:
    """Telegram notification handler"""
    
    def __init__(self):
        self.bot_token = os.getenv('TELEGRAM_BOT_TOKEN')
        self.chat_id = os.getenv('TELEGRAM_CHAT_ID')
        self.enabled = os.getenv('TELEGRAM_ENABLED', 'false').lower() == 'true'
    
    async def send_message(self, message: str):
        """Send message to Telegram"""
        if not self.enabled or not self.bot_token or not self.chat_id:
            print(f"📱 Telegram disabled or not configured")
            return False
        
        try:
            url = f"https://api.telegram.org/bot{self.bot_token}/sendMessage"
            async with aiohttp.ClientSession() as session:
                async with session.post(url, json={
                    'chat_id': self.chat_id,
                    'text': message,
                    'parse_mode': 'HTML'
                }) as response:
                    if response.status == 200:
                        print(f"✅ Telegram notification sent")
                        return True
                    else:
                        print(f"❌ Telegram failed: {response.status}")
                        return False
        except Exception as e:
            print(f"❌ Telegram error: {e}")
            return False

File: plugins/hyper-pnl-bot/src/telegram.js
import fetch from 'node-fetch';
import config from './config.js';

export async function sendReport(rep) {
  const getPnlEmoji = (val) => val >= 0 ? "🟢" : "🔴";
  const fmt = (val) => {
    const num = Number(val) || 0;
    return `${num >= 0 ? '+' : ''}${num.toFixed(2)}`;
  };

  let coinStr = "";
  for (const c in rep.byCoin) {
    const val = rep.byCoin[c];
    coinStr += `• ${c}: <code>${fmt(val)}</code> USDC\n`;
  }

  const pnlEmoji = getPnlEmoji(rep.net);
  const realizedEmoji = getPnlEmoji(rep.realized);
  const feeVal = Number(rep.fee) || 0;

  const text =
    `<b>📊 Hyperliquid PnL Report</b>
<code>${rep.wallet}</code>

<b>🏆 Win Rate: ${rep.winrate}%</b>
• Trades: <b>${rep.trades}</b>
• W/L: ${rep.wins}W - ${rep.losses}L

<b>💰 PnL Summary</b>
• Realized: <b>${fmt(rep.realized)} USDC</b> ${realizedEmoji}
• Fees: <code>-${feeVal.toFixed(2)} USDC</code>
• <b>Net PnL: ${fmt(rep.net)} USDC</b> ${pnlEmoji}

<b>💎 By Coin</b>
${coinStr || "No trades found."}`;

  const { token, chatId } = config.telegram;

  const res = await fetch(`https://api.telegram.org/bot${token}/sendMessage`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      chat_id: chatId,
      text,
      parse_mode: "HTML",
      disable_web_page_preview: true
    })
  });

  if (!res.ok) {
    const errorData = await res.text();
    console.error(`Telegram API Error: ${res.status} ${res.statusText}`, errorData);
    throw new Error(`Telegram API failed: ${errorData}`);
  } else {
    console.log(`Telegram report sent for ${rep.wallet}`);
  }
}

export async function sendHappyAlert(rep, threshold) {
  const { token, chatId } = config.telegram;
  const pnl = Number(rep.net) || 0;

  if (pnl <= threshold) return;

  const text =
    `🚀 <b>BIG WIN ALERT!</b> 🚀
<code>${rep.wallet}</code>

🎉 PnL (24h): <b>${pnl >= 0 ? '+' : ''}${pnl.toFixed(2)} USDC</b>
✅ Target: > ${threshold} USDC

Keep it up! 💰💰💰`;

  try {
    const res = await fetch(`https://api.telegram.org/bot${token}/sendMessage`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        chat_id: chatId,
        text,
        parse_mode: "HTML"
      })
    });

    if (res.ok) console.log(`Happy Alert sent for ${rep.wallet} (+${pnl} USDC)`);
  } catch (e) {
    console.error("Failed to send Happy Alert:", e);
  }
}
